tt_metal/hw/inc/debug/stack_usage.h

```diff
@@ -0,0 +1,89 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include "watcher_common.h"
+
+// We don't control the stack size for active erisc, and share the stack with base FW, so don't implement for ERISC.
+#if defined(WATCHER_ENABLED) && !defined(WATCHER_DISABLE_STACK_USAGE) && !defined(COMPILE_FOR_ERISC)
+
+#define STACK_DIRTY_PATTERN 0xBABABABA
+
+uint32_t get_stack_size() {
+#if defined(COMPILE_FOR_BRISC)
+    return MEM_BRISC_STACK_SIZE;
+#elif defined(COMPILE_FOR_NCRISC)
+    return MEM_NCRISC_STACK_SIZE;
+#elif defined(COMPILE_FOR_IDLE_ERISC)
+    return MEM_IERISC_STACK_SIZE;
+#elif COMPILE_FOR_TRISC == 0
+    return MEM_TRISC0_STACK_SIZE;
+#elif COMPILE_FOR_TRISC == 1
+    return MEM_TRISC1_STACK_SIZE;
+#elif COMPILE_FOR_TRISC == 2
+    return MEM_TRISC2_STACK_SIZE;
+#else
+    return 0;
+#endif
+}
+
+uint32_t get_stack_base() {
+#if defined(COMPILE_FOR_BRISC)
+    return MEM_BRISC_STACK_BASE;
+#elif defined(COMPILE_FOR_NCRISC)
+    return MEM_NCRISC_STACK_BASE;
+#elif defined(COMPILE_FOR_IDLE_ERISC)
+    return MEM_IERISC_STACK_BASE;
+#elif COMPILE_FOR_TRISC == 0
+    return MEM_TRISC0_STACK_BASE;
+#elif COMPILE_FOR_TRISC == 1
+    return MEM_TRISC1_STACK_BASE;
+#elif COMPILE_FOR_TRISC == 2
+    return MEM_TRISC2_STACK_BASE;
+#else
+    return 0;
+#endif
+}
+
+void dirty_stack_memory() {
+    uint32_t tt_l1_ptr *stack_ptr = (uint32_t tt_l1_ptr *) get_stack_base();
+    uint32_t stack_size = get_stack_size();
+    uint32_t stack_max_offset = stack_size / sizeof(uint32_t);
+    uint32_t stack_dirty_offset = stack_max_offset / 4;
+
+    // Dirty the back 3/4 of the stack (does rely on the fact that we haven't hit this part of the stack yet).
+    for (uint32_t stack_offset = stack_dirty_offset; stack_offset < stack_max_offset; stack_offset++) {
+        stack_ptr[stack_offset] = STACK_DIRTY_PATTERN;
+    }
+    return;
+}
+
+void record_stack_usage() {
+    // Write the pause flag for this core into the memory mailbox for host to read.
+    debug_stack_usage_t tt_l1_ptr *stack_usage_msg = GET_MAILBOX_ADDRESS_DEV(watcher.stack_usage);
+
+    uint32_t tt_l1_ptr *stack_ptr = (uint32_t tt_l1_ptr *) get_stack_base();
+    for (uint32_t stack_offset = 0; stack_offset < (get_stack_size() / sizeof(uint32_t)); stack_offset++) {
```

Comment:
confused, isn't this backwards?  stack grows high to low, you are starting low and going high, so we expect the first value to be the dirty pattern and we exit when we find not the dirty pattern ?

Reply:
Should be fixed now